src/utils/power_flow/power_flow.jl

```diff
@@ -139,12 +156,121 @@ solve_powerflow!(sys, nlsolve, method = :newton)
 ```
 
 """
-function solve_powerflow!(sys, nlsolve; args...)
-    pf!, x0 = PowerSystems.make_pf(sys)
+function solve_powerflow!(system::System, nlsolve; args...)
+    buses = sort(collect(get_components(Bus, system)), by = x -> get_number(x))
+    N_BUS = length(buses)
+
+    # assumes the ordering in YBus is the same as in the buses.
+    tempYb = Ybus(get_components(ACBranch, system), buses)
+    a = collect(1:N_BUS)
+    Yb = Ybus(tempYb.data, (a, a), (_make_ax_ref(a), _make_ax_ref(a))).data
+    I, J, V = SparseArrays.findnz(Yb)
+    neighbors = [Set{Int}([i]) for i in 1:N_BUS]
+    for nz in eachindex(V)
+        push!(neighbors[I[nz]], J[nz])
+        push!(neighbors[J[nz]], I[nz])
+    end
+    x0 = zeros(N_BUS * 2)
+
+    # Use vectors to cache data for closure
+    # These should be read only
+    P_GEN_BUS = fill(0.0, N_BUS)
+    Q_GEN_BUS = fill(0.0, N_BUS)
+    P_LOAD_BUS = fill(0.0, N_BUS)
+    Q_LOAD_BUS = fill(0.0, N_BUS)
+    P_net = fill(0.0, N_BUS)
+    Q_net = fill(0.0, N_BUS)
+    Vm = fill(0.0, N_BUS)
+    θ = fill(0.0, N_BUS)
+
+    state_variable_count = 1
+    sources = get_components(StaticInjection, system, d -> !isa(d, ElectricLoad))
+
+    for (ix, b) in enumerate(buses)
+        bus_number = get_number(b)::Int
+        bus_angle = get_angle(b)::Float64
+        bus_voltage = get_voltage(b)::Float64
+        P_GEN_BUS[ix] = 0.0
+        Q_GEN_BUS[ix] = 0.0
+        get_ext(b)["neighbors"] = neighbors[ix]
+        for gen in sources
+            if gen.bus == b
+                P_GEN_BUS[ix] += get_activepower(gen)
+                Q_GEN_BUS[ix] += get_reactivepower(gen)
+            end
+        end
+
+        P_LOAD_BUS[ix], Q_LOAD_BUS[ix] = _get_load_data(system, b)
+
+        if b.bustype == BusTypes.REF
+            injection_components = get_components(StaticInjection, system, d -> d.bus == b)
```

Comment:
A potential optimization is to create a `set` of buses attached to StaticInjection devices ahead of time.  This will iterate over the same components many times.

Reply:
There are other improvements missing that we can add in a second round. This will speed up the process of setting up the problem which is not currently a big bottle neck.